edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs

```diff
@@ -32,102 +33,167 @@ public async Task GetDevicesAndModulesInTargetDeviceScope([FromRoute] string act
             IList<ServiceIdentity> identities = await edgeHub.GetDevicesAndModulesInTargetScopeAsync(targetDeviceId);
+            Option<ServiceIdentity> targetDevice = await edgeHub.GetIdentityAsync(targetDeviceId);
+            targetDevice.ForEach(d => identities.Add(d));
+
+            // Construct the result from the identities
             Events.SendingScopeResult(identities);
             result = MakeResultFromIdentities(identities);
             result.Status = HttpStatusCode.OK;
 
             return result;
         }
 
-        async Task SendResponse(EdgeHubScopeResult result)
+        async Task<EdgeHubScopeResult> HandleGetDeviceAndModuleOnBehalfOf(string actorDeviceId, string actorModuleId, IdentityOnBehalfOfRequest request)
         {
-            this.Response.StatusCode = (int)result.Status;
+            Events.ReceivedIdentityOnBehalfOfRequest(actorDeviceId, request);
+            Preconditions.CheckNonWhiteSpace(request.TargetDeviceId, nameof(request.TargetDeviceId));
 
-            if (result.Status == HttpStatusCode.OK)
+            EdgeHubScopeResult result = new EdgeHubScopeResult();
+
+            bool isModule = false;
+            string targetId = request.TargetDeviceId;
+            if (!request.TargetModuleId.IsNullOrWhiteSpace())
             {
-                var resultJsonContent = JsonConvert.SerializeObject(result);
-                var resultUtf8Bytes = Encoding.UTF8.GetBytes(resultJsonContent);
+                isModule = true;
+                targetId += "/" + request.TargetModuleId;
+            }
 
-                this.Response.ContentLength = resultUtf8Bytes.Length;
-                this.Response.ContentType = SupportedContentType;
+            // Check that the actor device is authorized to act OnBehalfOf the target
+            IEdgeHub edgeHub = await this.edgeHubGetter;
+            if (!await this.AuthorizeActor(edgeHub, actorDeviceId, actorModuleId, targetId))
+            {
+                Events.UnauthorizedActor(actorDeviceId, actorModuleId, targetId);
+                result.Status = HttpStatusCode.Unauthorized;
+                return result;
+            }
 
-                await this.Response.Body.WriteAsync(resultUtf8Bytes, 0, resultUtf8Bytes.Length);
+            // Get the target identity
+            var identityList = new List<ServiceIdentity>();
+            Option<ServiceIdentity> identity = await edgeHub.GetIdentityAsync(targetId);
+            identity.ForEach(i => identityList.Add(i));
+
+            // If the target is a module, we also need to
+            // include the parent device as well to match
+            // IoT Hub API behavior
+            if (isModule)
+            {
+                Option<ServiceIdentity> device = await edgeHub.GetIdentityAsync(request.TargetDeviceId);
+                device.ForEach(i => identityList.Add(i));
             }
+
+            Events.SendingScopeResult(identityList);
```

Comment:
do we need to log scope result with inputs e.g. actor device id, actor module id, and IdentityOnBehalfOfRequest?

Reply:
actor is probably extraneous, considering that we're logging it at the beginning of this method, but the target id can be helpful, I'll add it.

---
In reply to: [458528369](https://github.com/Azure/iotedge/pull/3248#discussion_r458528369) [](ancestors = 458528369)